activerecord/lib/active_record/relation/where_clause_factory.rb

```diff
@@ -11,25 +127,36 @@ def initialize(klass, predicate_builder)
       def build(opts, other)
         case opts
         when String, Array
-          parts = [klass.sanitize_sql(other.empty? ? opts : ([opts] + other))]
+          all_opts = other.empty? ? opts : ([opts] + other)
+          clause = string_or_array_clause(all_opts)
         when Hash
-          attributes = predicate_builder.resolve_column_aliases(opts)
-          attributes = klass.send(:expand_hash_conditions_for_aggregates, attributes)
-          attributes.stringify_keys!
-
-          parts = predicate_builder.build_from_hash(attributes)
+          clause = HashClause.new(opts, @klass, @predicate_builder)
         when Arel::Nodes::Node
-          parts = [opts]
+          clause = ArelClause.new(opts)
         else
           raise ArgumentError, "Unsupported argument type: #{opts} (#{opts.class})"
         end
 
-        WhereClause.new(parts)
+        [WhereClause.new(clause.predicates), clause.references]
```

Comment:
Probably not ideal to return an array here. Could just return `clause` and build the `WhereClause` in `query_methods`, but then this class wouldn't be called `WhereClauseFactory` anymore. Suggestions are most welcome!